web/static/js/components/questionnaires/QuestionnaireEditor.jsx

```diff
@@ -39,7 +39,45 @@ class QuestionnaireEditor extends Component {
   }
 
   questionnaireAddStep(stepType) {
-    this.props.actions.addStep(stepType)
+    this.setState({
+      ...this.state,
+      addingStep: true
+    }, () => {
+      // Add the step then automatically expand it
+      this.props.questionnaireActions.addStep(stepType)
+    })
+  }
+
+  deleteStep() {
+    const currentStepId = this.state.currentStep
+
+    this.setState({
+      currentStep: null,
+      addingStep: false
+    }, () => {
+      this.props.questionnaireActions.deleteStep(currentStepId)
+    })
+  }
+
+  internalState(currentStep, addingStep = false) {
+    return {
+      currentStep,
+      addingStep
+    }
+  }
+
+  componentWillReceiveProps(newProps) {
+    // This feels a bit hacky, but it let's us expand the step we just created.
```

Comment:
It felt wrong because we'd have to add ad-hoc questionnaire substate to the store for the editor. This way we keep the reducer pretty focused on data management, and the editor manages UI state. 

But maybe it's not worth the effort, since as you said the previous approach seemed simpler. 

The questionnaire editor will suffer a lot of changes in the short term, so maybe we can move this to it's own issue.